pkg/sched/sched.go

```diff
@@ -149,38 +236,90 @@ func (s *manager) Start() {
 }
 
 func (s *manager) scheduleTasks() {
-	for {
-		select {
-		case <-s.ticker.C:
-			now := time.Now()
-			s.Lock()
-			tasksReady := make([]*taskInfo, 0)
-			for e := s.tasks.Front(); e != nil; e = e.Next() {
-				t := e.Value.(*taskInfo)
-				t.lock.Lock()
-				if !t.enqueued &&
-					(now.Equal(t.runAt) || now.After(t.runAt)) {
-					tasksReady = append(tasksReady, t)
-					t.enqueued = true
-					if t.onlyOnce {
-						s.tasks.Remove(e)
-					}
+	histLastPrinted := time.Now()
+	prevTick := time.Now()
+	for range s.ticker.C {
+		now := time.Now()
+		s.tickIntervalHist.record(now.Sub(prevTick))
+		s.Lock()
+		tasksReady := make([]*taskInfo, 0)
+		elementsToRemove := make([]*list.Element, 0)
+		for e := s.tasks.Front(); e != nil; e = e.Next() {
+			t := e.Value.(*taskInfo)
+			t.lock.Lock()
+			if !t.enqueued && (now.Equal(t.runAt) || now.After(t.runAt)) {
+				tasksReady = append(tasksReady, t)
+				t.enqueued = true
+				if t.onlyOnce {
+					elementsToRemove = append(elementsToRemove, e)
 				}
-				t.lock.Unlock()
 			}
-			s.Unlock()
-			s.enqueuedTasksLock.Lock()
-			for _, t := range tasksReady {
-				s.enqueuedTasks.PushBack(t)
+			t.lock.Unlock()
+		}
+		for _, e := range elementsToRemove {
+			s.tasks.Remove(e)
+		}
+		s.Unlock()
+
+		var numEnqueued int
+		s.enqueuedTasksLock.Lock()
+		for _, t := range tasksReady {
+			s.taskScheduleHist.record(time.Since(t.runAt))
+			s.enqueuedTasks.PushBack(t)
+		}
+		s.cv.Broadcast()
+		numEnqueued = s.enqueuedTasks.Len()
+		s.enqueuedTasksLock.Unlock()
+		workersStarted, workersExited := s.addWorkersIfNeeded(numEnqueued)
+
+		// print stats
+		if time.Since(histLastPrinted) > statsLogDuration {
+			logrus.Infof("sched stats: workers: current=%v, started=%v, exited=%v",
+				workersStarted-workersExited, workersStarted, workersExited)
+			for _, hist := range []histogram{
+				s.taskDurationHist, s.taskStartHist, s.taskScheduleHist, s.tickIntervalHist, s.tickProcessHist} {
+
+				logrus.Infof("sched stats: %s", hist.getHistogram())
 			}
-			s.cv.Broadcast()
-			s.enqueuedTasksLock.Unlock()
+			histLastPrinted = time.Now()
 		}
+		s.tickProcessHist.record(time.Since(now))
+		prevTick = time.Now()
 	}
 }
 
-func (s *manager) runTasks() {
+// for testing
+func (s *manager) getWorkerCount() int {
+	var ret int
+	s.Lock()
+	ret = int(s.workersStarted - s.workersExited)
```

Comment:
nit: can simply be `return int(s.workersStarted - s.workersExited)` with lock and defer ?